src/notifications.c

```diff
@@ -31,16 +44,70 @@ int HashNotificationCallback(RedisModuleCtx *ctx, int type, const char *event, R
 
   const char *key_cp = RedisModule_StringPtrLen(key, NULL);
   if (hset || hmset || hdel) {
-    Indexes_UpdateMatchingWithSchemaRules(ctx, key);
+    Indexes_UpdateMatchingWithSchemaRules(ctx, key, hashFields);
   }
   if (del) {
-    Indexes_DeleteMatchingWithSchemaRules(ctx, key);
+    Indexes_DeleteMatchingWithSchemaRules(ctx, key, hashFields);
   }
 
+  freeHashFields();
+
   return REDISMODULE_OK;
 }
 
+/*****************************************************************************/
+
+void CommandFilterCallback(RedisModuleCommandFilterCtx *filter) {
+  int numArgs = RedisModule_CommandFilterArgsCount(filter);
+  if (numArgs < 3) return;
+
+  bool hset = false;
+
+  size_t len;
+  const RedisModuleString *cmd = RedisModule_CommandFilterArgGet(filter, 0);
+  const char *cmdStr = RedisModule_StringPtrLen(cmd, &len);
+
+  if (!strcasecmp("HSET", cmdStr) || !strcasecmp("HMSET", cmdStr)) {
+    if (numArgs % 2 != 0) return;
+    hset = true;
+  } else if (!strcasecmp("HDEL", cmdStr)) {
+    // Nothing to do
+  } else {
+    return;
+  }
+  freeHashFields();
+
+  const RedisModuleString *keyStr = RedisModule_CommandFilterArgGet(filter, 1);
+  RedisModuleString *copyKeyStr = RedisModule_CreateStringFromString(RSDummyContext, keyStr);
+
+  RedisModuleKey *k = RedisModule_OpenKey(RSDummyContext, copyKeyStr, REDISMODULE_READ);
+  if (!k || RedisModule_KeyType(k) != REDISMODULE_KEYTYPE_HASH) {
+    // key does not exist or is not a hash, nothing to do
+    goto done;
+  }
+
+
+  // HSET receives field&value, HDEL receives field
+  int cmdFactor = hset ? 2 : 1;
+  int fieldsNum = (numArgs - 2) / cmdFactor;
+  hashFields = (char **)rm_calloc(fieldsNum + 1, sizeof(*hashFields));
+
+  for (size_t i = 0; i < fieldsNum; ++i) {
+    const RedisModuleString *field = RedisModule_CommandFilterArgGet(filter, 2 + i * cmdFactor);
+    const char *fieldStr = RedisModule_StringPtrLen(field, &len);
+    hashFields[i] = rm_strdup(fieldStr);

```

Comment:
Can we retain the string instead of copy it?

Reply:
@MeirShpilraien 
Tried retaining string. It crashes.
Possibly due to `robj` being on the stack.